### PR TITLE
fix(caldav): fixed initial sync and double processing

### LINKED ISSUE #44127: [Bug]: Deleted calendar objects included in initial sync
### ⚠️ This issue respects the following points: ⚠️

- [X] This is a **bug**, not a question or a configuration/webserver/proxy issue.
- [X] This issue is **not** already reported on [Github](https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3Abug) OR [Nextcloud Community Forum](https://help.nextcloud.com/) _(I've searched it)_.
- [X] Nextcloud Server **is** up to date. See [Maintenance and Release Schedule](https://github.com/nextcloud/server/wiki/Maintenance-and-Release-Schedule) for supported versions.
- [X] I agree to follow Nextcloud's [Code of Conduct](https://nextcloud.com/contribute/code-of-conduct/).

### Bug description

https://github.com/nextcloud/server/blob/f4beb95cdc3161a99e048e03ab486e5218d10e0a/apps/dav/lib/CalDAV/CalDavBackend.php#L2419-L2430 returns all object URIs regardless of the deleted_at flag. If a client runs into this, they might see download deleted events.

### Steps to reproduce

1. Have an existing calendar with deleted events
2. Connect a new caldav client
3. Synchronize

### Expected behavior

No deleted events synchronized

### Installation method

None

### Nextcloud Server version

master

### Operating system

None

### PHP engine version

None

### Web server

None

### Database engine version

None

### Is this bug present after an update or on a fresh install?

None

### Are you using the Nextcloud Server Encryption module?

None

### What user-backends are you using?

- [ ] Default user-backend _(database)_
- [ ] LDAP/ Active Directory
- [ ] SSO - SAML
- [ ] Other

### Configuration report

_No response_

### List of activated Apps

_No response_

### Nextcloud Signing status

_No response_

### Nextcloud Logs

_No response_

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by SebastianKrupinski**
* Resolves: #44127

## Summary
- Initial sync no longer returns deleted items
- fixed double processing of results when doing a follow up sync

### CODE REVIEW

File: apps/dav/lib/CalDAV/CalDavBackend.php [hunk 1]
```diff
@@ -2406,74 +2407,54 @@ public function getChangesForCalendar($calendarId, $syncToken, $syncLevel, $limi
...
-							break;
-						case 3:
-							$result['deleted'][] = $uri;
-							break;
-					}
-				}
+					->where($qb->expr()->eq('calendarid', $qb->createNamedParameter($calendarId)))
+					->andWhere($qb->expr()->eq('calendartype', $qb->createNamedParameter($calendarType)))
+					->groupBy('uri')
+					->having(new Literal('MAX(operation) < 3'));
```
Comment (ChristophWurst) [inline apps/dav/lib/CalDAV/CalDavBackend.php:2424]:
```
$qb->expr()->lt(
  $qb->func()->max('operation'),
  $qb->createNamedParameter(3),
)
```

**Review by ChristophWurst — CHANGES_REQUESTED** [review]
* Always use prepared statements with parameters
* Add a test to ``\OCA\DAV\Tests\unit\CalDAV\CalDavBackendTest``

File: apps/dav/lib/CalDAV/CalDavBackend.php (hunk 1 above)
Comment (SebastianKrupinski) [reply apps/dav/lib/CalDAV/CalDavBackend.php:2424]: Done, I've replaced the Literal with a prepared statement. 

As for the tests OCA\DAV\Tests\unit\CalDAV\CalDavBackendTest already has tests for the getChangesForCalendar function. Its tested 6 times.

Comment (ChristophWurst) [reply apps/dav/lib/CalDAV/CalDavBackend.php]: > As for the tests OCA\DAV\Tests\unit\CalDAV\CalDavBackendTest already has tests for the getChangesForCalendar function. Its tested 6 times.

I mean a test specifically for the initial sync and whether it includes deleted items

Comment (SebastianKrupinski) [reply apps/dav/lib/CalDAV/CalDavBackend.php]: Done

File: apps/dav/lib/CalDAV/CalDavBackend.php [hunk 2]
```diff
@@ -2410,74 +2410,59 @@ public function getChangesForCalendar($calendarId, $syncToken, $syncLevel, $limi
...
-							break;
-						case 3:
-							$result['deleted'][] = $uri;
-							break;
-					}
-				}
+					->where($qb->expr()->eq('calendarid', $qb->createNamedParameter($calendarId)))
+					->andWhere($qb->expr()->eq('calendartype', $qb->createNamedParameter($calendarType)))
+					->groupBy('uri')
+					// forced (PARAM_INT) is needed for doctriane with Sqlite, for the value to be used as an intager without quotes
```
Comment (st3iny) [inline apps/dav/lib/CalDAV/CalDavBackend.php:2427]:
```suggestion
					// forced (PARAM_INT) is needed for doctrine with Sqlite, for the value to be used as an integer without quotes
```

File: apps/dav/lib/CalDAV/CalDavBackend.php [hunk 3]
```diff
@@ -2410,74 +2410,59 @@ public function getChangesForCalendar($calendarId, $syncToken, $syncLevel, $limi
...
+					->having($qb->expr()->lt($qb->func()->max('operation'), $qb->createNamedParameter(3, $qb::PARAM_INT)));
```
Comment (st3iny) [inline apps/dav/lib/CalDAV/CalDavBackend.php:2428]: Magic number 3? What does 3 mean in this context?

**Review by st3iny — COMMENTED** [review]
This is also a (minor) performance optimization as we do `$qb->func()->max('operation')` instead of a busy loop now :+1: 

I'm gonna give this a test but looks good so far.

File: apps/dav/tests/unit/CalDAV/AbstractCalDavBackend.php [hunk 4]
```diff
@@ -207,6 +207,36 @@ protected function createEvent($calendarId, $start = '20130912T130000Z', $end =
...
+SUMMARY:Test Event
+DTSTART;VALUE=DATE-TIME:$start
+DTEND;VALUE=DATE-TIME:$end
+CLASS:PUBLIC
+END:VEVENT
+END:VCALENDAR
+EOD;
+
+		$this->backend->updateCalendarObject($calendarId, $objectId, $calData);
+
```
Comment (st3iny) [inline apps/dav/tests/unit/CalDAV/AbstractCalDavBackend.php:231]:
```suggestion
```

File: apps/dav/tests/unit/CalDAV/AbstractCalDavBackend.php [hunk 5]
```diff
@@ -207,6 +207,36 @@ protected function createEvent($calendarId, $start = '20130912T130000Z', $end =
...
+	}
+
+	protected function deleteEvent($calendarId, $objectId) {
+
```
Comment (st3iny) [inline apps/dav/tests/unit/CalDAV/AbstractCalDavBackend.php:235]:
```suggestion
```

File: apps/dav/tests/unit/CalDAV/AbstractCalDavBackend.php [hunk 6]
```diff
@@ -207,6 +207,36 @@ protected function createEvent($calendarId, $start = '20130912T130000Z', $end =
...
+		$this->backend->deleteCalendarObject($calendarId, $objectId);
+
```
Comment (st3iny) [inline apps/dav/tests/unit/CalDAV/AbstractCalDavBackend.php:237]:
```suggestion
```

File: apps/dav/lib/CalDAV/CalDavBackend.php (hunk 3 above)
Comment (SebastianKrupinski) [reply apps/dav/lib/CalDAV/CalDavBackend.php:2428]: The magic number is the deleted item operation value. There is no constant defined in the class it's hard coded every where.

File: apps/dav/lib/CalDAV/CalDavBackend.php (hunk 2 above)
Comment (SebastianKrupinski) [reply apps/dav/lib/CalDAV/CalDavBackend.php:2427]: Done

File: apps/dav/tests/unit/CalDAV/AbstractCalDavBackend.php (hunk 6 above)
Comment (SebastianKrupinski) [reply apps/dav/tests/unit/CalDAV/AbstractCalDavBackend.php:237]: Done

File: apps/dav/tests/unit/CalDAV/CalDavBackendTest.php [hunk 7]
```diff
@@ -468,19 +468,91 @@ public function providesCalendarQueryParameters() {
 		];
 	}
 
-	public function testSyncSupport(): void {
-		$calendarId = $this->createTestCalendar();
+	public function testCalendarSynchronization(): void {
```
[the remark this answers is not in the document]
Comment (SebastianKrupinski) [reply apps/dav/tests/unit/CalDAV/CalDavBackendTest.php:471]: I agree making it easier to see what failed is very helpful. But I think using a custom assert message might be better, that way we don't need to setup multiple tests that do basically the same thing and can group things more logically, less code and less testing time.

How is this for a failure message? The second line.

```
1) OCA\DAV\Tests\unit\CalDAV\CalDavBackendTest::testCalendarSynchronization
Failed test delta sync state with modified/deleted events in calendar
Failed asserting that two arrays are equal.
--- Expected
+++ Actual
@@ @@
 Array (
     'syncToken' => 6
     'added' => Array (
-        0 => 'eventVnNOXM4Zb6xvK'
     )
     'modified' => Array (
-        0 => 'eventxuOQWjKl5G5JQ'
     )
     'deleted' => Array (
-        0 => 'eventsQDXnXVcvfN3M'
     )
 )
```

[the remark this answers is not in the document]
Comment (st3iny) [reply apps/dav/tests/unit/CalDAV/CalDavBackendTest.php:471]: I like the approach :+1:

File: apps/dav/lib/CalDAV/CalDavBackend.php [hunk 8]
```diff
@@ -2410,74 +2410,59 @@ public function getChangesForCalendar($calendarId, $syncToken, $syncLevel, $limi
...
-			];
-
-			if ($syncToken) {
+			// evaluate if this is a initial sync and construct appropriate command
+			if ($initialSync) {
 				$qb = $this->db->getQueryBuilder();
-
-				$qb->select('uri', 'operation')
+				$qb->select('uri')
 					->from('calendarchanges')
```
Comment (st3iny) [inline apps/dav/lib/CalDAV/CalDavBackend.php:2423]: This seems dangerous to me. Why do we pull calendar changes on the initial sync?

The problem is that calendar changes are periodically trimmed so the changes table might not contain data about all events.

Shouldn't the initial sync fetch all events from the `calendarobjects` table like before?

**Review by st3iny — CHANGES_REQUESTED** [review]

**Review by miaulalala — APPROVED** [review]

File: apps/dav/lib/CalDAV/CalDavBackend.php (hunk 8 above)
Comment (SebastianKrupinski) [reply apps/dav/lib/CalDAV/CalDavBackend.php:2423]: As per our call... (face palm)... did realize that the operations just get truncated every few months... They should be selectively pruned to only include the last operation for each object...  Adjusted the code to accommodate truncation...

File: apps/dav/tests/unit/CalDAV/CalDavBackendTest.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (SebastianKrupinski) [reply apps/dav/tests/unit/CalDAV/CalDavBackendTest.php:471]: Perfect

**Review by st3iny — APPROVED** [review]
LGTM! Refactoring of the incremental sync case is a (minor) performance improvement. Additional unit tests are always welcome.

**Review by ChristophWurst — APPROVED** [review]

**Pull request merged**